File: csharp/src/Audio/OmniGain.cs
using System;

namespace BQt.Audio
{
    public class OmniGain : IAudioNode
    {
        private float _volume = 1.0f;
        private readonly object _lock = new object();

        public string Name => "OmniGain";

        public void SetVolume(float vol)
        {
            lock (_lock)
            {
                _volume = vol;
            }
        }

        public void Process(float[] buffer)
        {
            lock (_lock)
            {
                if (Math.Abs(_volume - 1.0f) < float.Epsilon)
                {
                    return;
                }

                for (int i = 0; i < buffer.Length; i++)
                {
                    buffer[i] *= _volume;
                }
            }
        }
    }
}

File: rust/src/audio/omni_gain.rs
use std::sync::{Arc, Mutex};
use super::audio_graph::AudioNode;

pub struct OmniGain {
    volume: Arc<Mutex<f32>>,
}

impl OmniGain {
    pub fn new() -> Self {
        Self {
            volume: Arc::new(Mutex::new(1.0)),
        }
    }

    pub fn set_volume(&self, vol: f32) {
        let mut v = self.volume.lock().unwrap();
        *v = vol;
    }
}

impl AudioNode for OmniGain {
    fn name(&self) -> &str {
        "OmniGain"
    }

    fn process(&mut self, buffer: &mut [f32]) {
        let vol = *self.volume.lock().unwrap();
        if (vol - 1.0).abs() < f32::EPSILON {
            return;
        }

        for sample in buffer.iter_mut() {
            *sample *= vol;
        }
    }
}
